src/main/java/org/jnode/fs/xfs/XfsObject.java

```diff
@@ -97,4 +123,72 @@ public long getUInt32(int relativeOffset) {
     public long getInt64(int relativeOffset) {
         return BigEndian.getInt64(data, offset + relativeOffset);
     }
+
+    public long read(long offset, int size) throws IOException {
+        switch (size) {
+            case 1:
+                return getUInt8((int) offset);
+            case 2:
+                return getUInt16((int) offset);
+            case 3:
+                return getUInt24((int) offset);
+            case 4:
+                return getUInt32((int) offset);
+            case 6:
+                return getUInt48((int) offset);
+            case 8:
+                return getInt64((int) offset);
+        }
+        throw new RuntimeException("Invalid read size " + size);
+    }
+
+    protected String readAsHexString(long offset, int size) throws IOException {
+        final long data = read(offset, size);
+        return Long.toHexString(data);
+    }
+
+    protected String readUuid(long offset, int size) throws IOException {
+                 return readAsHexString(offset, 4)
+                + "-" + readAsHexString(offset + 4, 2)
+                + "-" + readAsHexString(offset + 6, 2)
+                + "-" + readAsHexString(offset + 8, 2)
+                + "-" + readAsHexString(offset + 10, 6);
+    }
+
+
+    public static long asciiToHex(String asciiString) {
+        StringBuilder hex = new StringBuilder();
+        for (char c : asciiString.toCharArray()) {
+            hex.append(Integer.toHexString(c).toUpperCase(Locale.ROOT));
+        }
+        return Long.parseLong(hex.toString(), 16);
+    }
+
+    public static String hexToAscii(String hexString) {
+        try {
+            StringBuilder ascii = new StringBuilder();
+            final Iterable<String> chars = Splitter.fixedLength(2).split(hexString);
+            for (String c : chars) {
+                ascii.append((char) Byte.parseByte(c, 16));
+            }
+            return ascii.toString();
+        } catch (Throwable t) {
+            return "INVALID";
+        }
+    }
+
+    abstract protected List<Long> validSignatures();
+
+    public boolean isValidSignature() throws IOException{
```

Comment:
The whole inheritance hierarchy around this seems unnecessary. The way this is being done means that `validSignatures()` needs to be implemented by all subclasses - even those that don't need or use it, and `getMagicSignature()` has a default implementation which is then overridden by sublcasses to a default value of 0 when it's not being used. This would probably be easier to read if this was implemented in the locations its currently being used in (LeafInfo, BPlusTreeDataExtent, and INode) and the `XfsObject` implementation is removed.

Reply:
Code changes were applied removing the class and the methods that were not required.